R/methods-Module.R

```diff
@@ -0,0 +1,767 @@
+  hotspot <- import("hotspot", convert=F)
+  pickle <- import("pickle", convert=F)
+  py$pickle <- pickle
+  
+  if (!is.null(file)) {
+    # load file
+    hs <- py_load_object(file)
+    return(hs)
+  } else if (!is.null(bytes)) {
+    py$hs_bytes <- r_to_py(bytes)
+    py_run_string("hs = pickle.loads(hs_bytes)")
+    return(py$hs)
+  } else {
+    return(NULL)
+  }
+}
+
+
+#' Save bytes in the Hotspot object slot to a file
+#' @param path the file path
+#' @param bytes the raw bytes, like in the Hotspot slot of a VISION Object
+#' 
+#' @export
+saveHSBytestToPickle <- function(path, bytes) {
+  py_save_object(obj=loadHotspotObject(bytes=bytes), path)
+}
+
+
+#' Add custom tree based neighbor and weights to a Hotspot object
+#' 
+#'  @param tree object of class phylo
+#'  @param the Hotspot object to add the nw to
+#'  @param minSize the minimum number of neighbors of the node
+#'  @return the Hotspot object
+#'  
+#'  @export
+lcaBasedHotspotNeighbors <- function(tree, hotspot, minSize=20) {
+  tips <- tree$tip.label
+  nTips <- length(tips)
+  neighbors <- data.frame(t(matrix(seq_len(nTips) -1, ncol = nTips, nrow= nTips)))
+  rownames(neighbors) <- tips
+  
+  
+  weights <- data.frame(matrix(0, ncol = nTips, nrow= nTips))
+  for (tip in seq_len(nTips)) {
+    my_neighbors <- minSizeCladeNeighbors(tree, tip, minSize)
+    
+    weights[tip, my_neighbors] <- 1
+  }
+  
+  neighbors_no_diag <- data.frame(matrix(ncol = nTips -1, nrow= nTips))
+  weights_no_diag <- data.frame(matrix(ncol = nTips -1, nrow= nTips))
+  
+  for (tip in seq_len(nTips)) {
+    neighbors_no_diag[tip, ] <- neighbors[tip, -tip]
+    weights_no_diag[tip, ] <- weights[tip, -tip]
+  }
+  
+  rownames(neighbors_no_diag) <- tips
+  rownames(weights_no_diag) <- tips
+  
+  colnames(neighbors_no_diag) <- seq_len(nTips-1) - 1
+  colnames(weights_no_diag) <- seq_len(nTips-1) - 1
+  return(list("neighbors"=neighbors_no_diag, "weights"=weights_no_diag))
+}
+
+
+#' Draw Modules Heatmap (Gene x Gene)
+#' 
+#' @param hs the Hotspot Object
+#' @param palette palette
+#' @export
+draw_hotspot_heatmap <- function(hs, palette = paletteer_d("ggsci::default_nejm")) {
+  scipy_hierarchy <- import("scipy.cluster.hierarchy", convert=F)
+  np <- import("numpy")
+  linkage <- hs$linkage
+  py_dend = scipy_hierarchy$dendrogram(linkage)
+  lcz = hs$local_correlation_z
+  gene_order = colnames(lcz)[np$array(py_dend$leaves)+1]
+  col_mapping = c()
+  module_to_col = list()
+  hs$modules = as.character(hs$modules)
+  unique_mods = as.character(unique(hs$modules))
+  example_genes = list()
+  col_mapping[["-1"]] = "#ffffff"
+  for (i in 1:length(unique_mods)) {
+    mod = unique_mods[[i]]
+    if (mod != -1) {
+      col_mapping[[mod]] = palette[[i]] 
+      module_to_col[[mod]] = palette[[i]]
+      example_genes[[mod]] = sample(hs$modules[hs$modules == mod], 5)
+    }
+  }
+  print(module_to_col)
+  print(col_mapping[order(names(col_mapping))])
+  print(col_mapping)
+  print(example_genes)
```

Comment:
We don't need these print statements anymore.

Also, did we ever end up making sure this worked in your hands? If I recall correctly, it wasn't working for you though it worked in my jupyter notebooks.